File: app/src/test/java/com/coinninja/coinkeeper/service/runner/FulfillSentInvitesRunnerTest.java
package com.coinninja.coinkeeper.service.runner;

import android.content.res.Resources;

import com.coinninja.bindings.TransactionBuilder;
import com.coinninja.coinkeeper.CoinKeeperApplication;
import com.coinninja.coinkeeper.model.PhoneNumber;
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary;
import com.coinninja.coinkeeper.model.helpers.TransactionHelper;

import org.junit.After;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import java.util.ArrayList;
import java.util.List;

import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class FulfillSentInvitesRunnerTest {

    @Mock
    BroadcastBtcInviteRunner broadcastBtcInviteRunner;
    @Mock
    private PhoneNumber receiverPhoneNumber;
    @Mock
    private PhoneNumber senderPhoneNumber;
    @Mock
    private CoinKeeperApplication application;
    @Mock
    private SentInvitesStatusGetter statusRunner;
    @Mock
    private TransactionHelper txHelper;
    @Mock
    private SentInvitesStatusSender sender;
    @InjectMocks
    private FulfillSentInvitesRunner fulfillRunner;

    @After
    public void tearDown() throws Exception {
        application = null;
        statusRunner = null;
        txHelper = null;
        sender = null;
        broadcastBtcInviteRunner = null;
        fulfillRunner = null;
        receiverPhoneNumber = null;
        senderPhoneNumber = null;
    }

    @Test
    public void successful_full_flow() {

        fulfillRunner.run();

        //Step 1. grab all sent invites from server, save/update the ones that now have an address
        verify(statusRunner).run();

        //Step 2. get any sent invites that do not have a tx id but have an address
        verify(txHelper).gatherUnfulfilledInviteTrans();

        //Step 3. report to coinninja server of any invites that have been newly fulfilled (with TX ID)
        verify(sender).run();
    }

    @Test
    public void one_unfulfilled_test() {
        InviteTransactionSummary unfulfilled = mock(InviteTransactionSummary.class);

        List<InviteTransactionSummary> unfulfilledTransactions = new ArrayList<>();
        unfulfilledTransactions.add(unfulfilled);
        TransactionBuilder txBuilder = mock(TransactionBuilder.class);
        when(unfulfilled.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);

        Resources re = mock(Resources.class);
        when(re.getString(anyInt())).thenReturn("Some String");
        when(application.getResources()).thenReturn(re);

        when(txHelper.gatherUnfulfilledInviteTrans()).thenReturn(unfulfilledTransactions);

        fulfillRunner.run();


        verify(broadcastBtcInviteRunner).run();
    }

    @Test
    public void three_or_more_unfulfilled_test() {
        InviteTransactionSummary unfulfilled1 = mock(InviteTransactionSummary.class);
        InviteTransactionSummary unfulfilled2 = mock(InviteTransactionSummary.class);
        InviteTransactionSummary unfulfilled3 = mock(InviteTransactionSummary.class);

        List<InviteTransactionSummary> unfulfilledTransactions = new ArrayList<>();
        unfulfilledTransactions.add(unfulfilled1);
        unfulfilledTransactions.add(unfulfilled2);
        unfulfilledTransactions.add(unfulfilled3);
        when(unfulfilled1.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled1.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);
        when(unfulfilled2.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled2.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);
        when(unfulfilled3.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled3.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);

        Resources re = mock(Resources.class);
        when(re.getString(anyInt())).thenReturn("Some String");
        when(application.getResources()).thenReturn(re);

        when(txHelper.gatherUnfulfilledInviteTrans()).thenReturn(unfulfilledTransactions);

        fulfillRunner.run();


        verify(broadcastBtcInviteRunner, times(3)).run();
    }

    @Test
    public void skip_invites_with_tx_id_test() {
        String sampleTxID = "some tx id";
        InviteTransactionSummary unfulfilled1 = mock(InviteTransactionSummary.class);
        InviteTransactionSummary fulfilled2 = mock(InviteTransactionSummary.class);
        InviteTransactionSummary unfulfilled3 = mock(InviteTransactionSummary.class);

        List<InviteTransactionSummary> unfulfilledTransactions = new ArrayList<>();
        unfulfilledTransactions.add(unfulfilled1);
        unfulfilledTransactions.add(fulfilled2);
        unfulfilledTransactions.add(unfulfilled3);
        when(unfulfilled1.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled1.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);
        when(fulfilled2.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(fulfilled2.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);
        when(fulfilled2.getBtcTransactionId()).thenReturn(sampleTxID);
        when(unfulfilled3.getReceiverPhoneNumber()).thenReturn(receiverPhoneNumber);
        when(unfulfilled3.getSenderPhoneNumber()).thenReturn(senderPhoneNumber);

        Resources re = mock(Resources.class);
        when(re.getString(anyInt())).thenReturn("Some String");
        when(application.getResources()).thenReturn(re);

        when(txHelper.gatherUnfulfilledInviteTrans()).thenReturn(unfulfilledTransactions);

        fulfillRunner.run();

        //only call two times, even tho we have 3 invitesop
        verify(broadcastBtcInviteRunner, times(2)).run();
    }

    @Test
    public void check_any_invite_to_see_if_it_has_a_tx_id_test() {
        String sampleTxID = "some tx id";
        InviteTransactionSummary fulfilled = mock(InviteTransactionSummary.class);
        when(fulfilled.getBtcTransactionId()).thenReturn(sampleTxID);

        boolean alreadyHasTxId = fulfillRunner.alreadyHasTxId(fulfilled);

        assertTrue(alreadyHasTxId);
    }

    @Test
    public void check_any_invite_to_see_if_it_has_a_null_tx_id_test() {
        String sampleTxID = null;
        InviteTransactionSummary fulfilled = mock(InviteTransactionSummary.class);
        when(fulfilled.getBtcTransactionId()).thenReturn(sampleTxID);

        boolean alreadyHasTxId = fulfillRunner.alreadyHasTxId(fulfilled);

        assertFalse(alreadyHasTxId);
    }

    @Test
    public void check_any_invite_to_see_if_it_has_a_empty_tx_id_test() {
        String sampleTxID = "";
        InviteTransactionSummary fulfilled = mock(InviteTransactionSummary.class);
        when(fulfilled.getBtcTransactionId()).thenReturn(sampleTxID);

        boolean alreadyHasTxId = fulfillRunner.alreadyHasTxId(fulfilled);

        assertFalse(alreadyHasTxId);
    }
}

File: app/src/test/java/com/coinninja/coinkeeper/service/runner/FulfillSentInvitesRunnerTest.kt
package com.coinninja.coinkeeper.service.runner

import app.coinninja.cn.thunderdome.model.LedgerInvoice
import com.coinninja.coinkeeper.model.db.InviteTransactionSummary
import com.coinninja.coinkeeper.model.db.enums.BTCState
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import junit.framework.Assert.assertFalse
import junit.framework.Assert.assertTrue
import org.junit.Test
import org.mockito.Mockito.*

class FulfillSentInvitesRunnerTest {

    private fun createRunner(): FulfillSentInvitesRunner =
            FulfillSentInvitesRunner(mock(), mock(), mock(), mock(), mock(), mock())

    @Test
    fun successful_full_flow() {
        val runner = createRunner()
        val invite = mock(InviteTransactionSummary::class.java)
        whenever(invite.btcTransactionId).thenReturn(null)
        whenever(runner.inviteTransactionSummaryHelper.unfulfilledSentInvites).thenReturn(listOf(invite))

        val ordered = inOrder(runner.inviteTransactionSummaryHelper, runner.sentInvitesStatusGetter, runner.sentInvitesStatusSender, runner.broadcastBtcInviteRunner)
        runner.run()

        //Step 1. grab all sent invites from server, save/update the ones that now have an address
        ordered.verify(runner.sentInvitesStatusGetter).run()

        //Step 2. addressForPubKey any sent invites that do not have a tx id but have an address
        ordered.verify(runner.inviteTransactionSummaryHelper).unfulfilledSentInvites

        //Step 3. report to coinninja server of any invites that have been newly fulfilled (with TX ID)
        ordered.verify(runner.broadcastBtcInviteRunner).invite = invite
        ordered.verify(runner.broadcastBtcInviteRunner).run()

        ordered.verify(runner.sentInvitesStatusSender).run()
    }

    @Test
    fun broadcast_invites_on_thunderdome() {
        val runner = createRunner()
        val pendingInvite: InviteTransactionSummary = mock()
        whenever(pendingInvite.serverId).thenReturn("--server-id--")
        val invite: InviteTransactionSummary = mock()
        whenever(invite.serverId).thenReturn("--server-id--")
        whenever(invite.address).thenReturn("--ln-invoice-id--")
        whenever(invite.valueSatoshis).thenReturn(100_000)
        whenever(runner.thunderDomeRepository.pay("--ln-invoice-id--", 100_000))
                .thenReturn(LedgerInvoice(value = 100_000, id = "--txid--"))

        runner.broadcastInvitesOnThunderDome(listOf(pendingInvite, invite))

        val ordered = inOrder(runner.thunderDomeRepository, invite, runner.cnClient)

        ordered.verify(runner.thunderDomeRepository).pay("--ln-invoice-id--", 100_000)
        ordered.verify(invite).btcState = BTCState.FULFILLED
        ordered.verify(invite).update()
        ordered.verify(runner.cnClient).updateInviteStatusCompleted("--server-id--", "--txid--")
    }

    @Test
    fun one_unfulfilled_test() {
        val runner = createRunner()
        val unfulfilled = mock(InviteTransactionSummary::class.java)
        val unfulfilledTransactions = listOf(unfulfilled)
        whenever(runner.inviteTransactionSummaryHelper.unfulfilledSentInvites).thenReturn(unfulfilledTransactions)

        runner.run()

        verify(runner.broadcastBtcInviteRunner).run()
    }

    @Test
    fun three_or_more_unfulfilled_test() {
        val runner = createRunner()
        val unfulfilled1 = mock(InviteTransactionSummary::class.java)
        val unfulfilled2 = mock(InviteTransactionSummary::class.java)
        val unfulfilled3 = mock(InviteTransactionSummary::class.java)
        val unfulfilledTransactions = listOf(unfulfilled1, unfulfilled2, unfulfilled3)
        whenever(runner.inviteTransactionSummaryHelper.unfulfilledSentInvites).thenReturn(unfulfilledTransactions)

        runner.run()

        verify(runner.broadcastBtcInviteRunner, times(3)).run()
    }

    @Test
    fun skip_invites_with_tx_id_test() {
        val runner = createRunner()
        val unfulfilled1 = mock(InviteTransactionSummary::class.java)
        val fulfilled2 = mock(InviteTransactionSummary::class.java)
        val unfulfilled3 = mock(InviteTransactionSummary::class.java)
        val unfulfilledTransactions = listOf(unfulfilled1, fulfilled2, unfulfilled3)
        whenever(fulfilled2.btcTransactionId).thenReturn("--txid--")
        whenever(runner.inviteTransactionSummaryHelper.unfulfilledSentInvites).thenReturn(unfulfilledTransactions)

        runner.run()

        verify(runner.broadcastBtcInviteRunner, times(2)).run()
    }

    @Test
    fun check_any_invite_to_see_if_it_has_a_tx_id_test() {
        val runner = createRunner()
        val sampleTxID = "some tx id"
        val fulfilled = mock(InviteTransactionSummary::class.java)
        whenever(fulfilled.btcTransactionId).thenReturn(sampleTxID)

        val alreadyHasTxId = runner.alreadyHasTxId(fulfilled)

        assertTrue(alreadyHasTxId)
    }

    @Test
    fun check_any_invite_to_see_if_it_has_a_null_tx_id_test() {
        val runner = createRunner()
        val sampleTxID: String? = null
        val fulfilled = mock(InviteTransactionSummary::class.java)
        whenever(fulfilled.btcTransactionId).thenReturn(sampleTxID)

        val alreadyHasTxId = runner.alreadyHasTxId(fulfilled)

        assertFalse(alreadyHasTxId)
    }

    @Test
    fun check_any_invite_to_see_if_it_has_a_empty_tx_id_test() {
        val runner = createRunner()
        val sampleTxID = ""
        val fulfilled = mock(InviteTransactionSummary::class.java)
        whenever(fulfilled.btcTransactionId).thenReturn(sampleTxID)

        val alreadyHasTxId = runner.alreadyHasTxId(fulfilled)

        assertFalse(alreadyHasTxId)
    }
}